brian2/core/clocks.py

```diff
@@ -61,8 +61,98 @@ def check_dt(new_dt, old_dt, target_t):
             f"time {t} is not a multiple of {new}."
         )
 
+class ClockArray:
+    def __init__(self, clock):
+        self.clock = clock
+    
+    def __getitem__(self, timestep):
+        return self.clock.dt * timestep
+    
+class EventClock(VariableOwner):
+    def __init__(self, times, name="eventclock*"):
+        Nameable.__init__(self, name=name)
+        self.variables = Variables(self)
+        self.times = times
+        self.variables.add_array(
+            "timestep", size=1, dtype=np.int64, read_only=True, scalar=True
+        )
+        self.variables.add_array(
+            "t",
+            dimensions=second.dim,
+            size=1,
+            dtype=np.float64,
+            read_only=True,
+            scalar=True,
+        )
+        self.variables["timestep"].set_value(0)
+        self.variables["t"].set_value(self.times[0])
+
+        self.variables.add_constant("N", value=1)
+
+        self._enable_group_attributes()
+        
+        self._i_end = None
+        logger.diagnostic(f"Created clock {self.name}")
+
+    def advance(self):
+        """
+        Advance the clock to the next timestep.
+        """
+        current_timestep = self.variables["timestep"].get_value().item()
+        next_timestep = current_timestep + 1
+        if self._i_end is not None and next_timestep > self._i_end:
+            raise StopIteration("Clock has reached the end of its available times.")
+        else:
+            self.variables["timestep"].set_value(next_timestep)
+            self.variables["t"].set_value(self.times[next_timestep])
+    
+    @check_units(start=second, end=second)
+    def set_interval(self, start, end):
+        """
+        Set the start and end time of the simulation.
+        """
+
+        if not isinstance(self.times, ClockArray):
 
-class Clock(VariableOwner):
+            start_idx = np.searchsorted(self.times, float(start))
+            end_idx = np.searchsorted(self.times, float(end))
+            
+            self.variables["timestep"].set_value(start_idx)
+            self.variables["t"].set_value(self.times[start_idx])
+            self._i_end = end_idx - 1 
+        else:
+
+            pass
+            
+    def __lt__(self, other):
+        return self.variables["t"].get_value().item() < other.variables["t"].get_value().item()
+
+    def __eq__(self, other):
```

Comment:
While I get the appeal of using `__eq__` here, I don't think it is the right approach – also, it breaks everything :) The reason is that we don't want to consider two `Clock`s to be equal just because their time is the same. E.g. there are places where we check whether two objects use the same clock, and those would always return `True` at the start of a simulation when those clocks are at 0ms. Also, we have places where we put clocks into a `set`, and Python objects that have `__eq__`  but not `__hash__` are not hashable (and the hash of an object is not allowed to change and needs to be consistent with `__eq__` so we cannot simply add a `__hash__` function. I think the best approach would be to add a specific function to compare them, say `same_time(self, other)`